### PR TITLE
Pull renderer out of xterm.js

### LINKED ISSUE #266: Allow xterm.js to exist and receive/send data before it's attached to the DOM
In v1.6 of VS Code the newly introduced Terminal API will allow processes to exist before the actual terminal panel/element is created. While this is currently possible, xterm cannot be initialized until the element exists, which means the output before the element is attached is lost. It would be good if everything was functional only no rendering took place until `open` is invoked.

### PR DESCRIPTION
**Pull request by Tyriar**
Fixes #266
Merge https://github.com/sourcelair/xterm.js/pull/521 first

---

This makes it possible to hook up messages with the pty and they are retained before `Terminal.open` is called.

### CODE REVIEW

File: src/Renderer.ts [hunk 1]
```diff
@@ -0,0 +1,303 @@
+import { ITerminal } from './Interfaces';
```
Comment (parisk) [inline src/Renderer.ts]: License comment? 😊

File: src/xterm.js [hunk 2]
```diff
@@ -1065,261 +1050,26 @@ Terminal.prototype.destroy = function() {
...
-        }
-      }
-      this.refreshRowsQueue = [];
-      this.refresh(start, end);
-    }
-  }
-  window.requestAnimationFrame(this.refreshLoop.bind(this));
-}
+Terminal.prototype.refresh = function(start, end) {
+  this.queueRefresh(start, end);
```
Comment (parisk) [inline src/xterm.js]: Should we call `this.renderer.queueRefresh` directly?

File: src/xterm.js [hunk 3]
```diff
@@ -1065,261 +1050,26 @@ Terminal.prototype.destroy = function() {
...
-    if (attr !== this.defAttr) {
-      out += '</span>';
-    }
-
-    this.children[y].innerHTML = out;
-  }
-
-  if (parent) {
-    this.element.appendChild(this.rowContainer);
+Terminal.prototype.queueRefresh = function(start, end) {
```
Comment (parisk) [inline src/xterm.js]: Do we need this method in the Terminal prototype?

I think we can "hide" from Terminal that queueing even exists.

Comment (Tyriar) [inline src/xterm.js]: This was a facade to reduce the diff, I'll clean it up.

File: src/xterm.js (hunk 2 above)
Comment (Tyriar) [inline src/xterm.js]: I like it, queueRefresh was a facade to simplify the change but there aren't as many calls as I expected so it's easy to change.

**Comment by Tyriar** [comment]
@parisk made the changes, did some clean up and fixed a nasty bug (renderer disappears after a reset). I'll merge this now :+1:

**Pull request merged**